File: android-pgs-plugin/src/SavedGamesController.java
package org.godotengine.godot;

import android.app.Activity;
import android.content.Intent;
import android.util.Pair;
import android.util.Log;

import androidx.annotation.NonNull;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.games.Games;
import com.google.android.gms.games.SnapshotsClient;
import com.google.android.gms.games.snapshot.Snapshot;
import com.google.android.gms.games.snapshot.SnapshotMetadata;
import com.google.android.gms.games.snapshot.SnapshotMetadataChange;
import com.google.android.gms.tasks.Continuation;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;

import java.io.IOException;

public class SavedGamesController {

    static final int RC_SAVED_GAMES = 9009;

    private Activity activity;
    private GodotCallbacksUtils godotCallbacksUtils;
    private ConnectionController connectionController;

    SavedGamesController(Activity activity, GodotCallbacksUtils godotCallbacksUtils, ConnectionController connectionController) {
        this.activity = activity;
        this.godotCallbacksUtils = godotCallbacksUtils;
        this.connectionController = connectionController;
    }

    protected void showSavedGamesUI(String title, boolean allowAddBtn, boolean allowDeleteBtn, int maxNumberOfSavedGamesToShow) {
        GoogleSignInAccount googleSignInAccount = GoogleSignIn.getLastSignedInAccount(activity);
        if (connectionController.isConnected().first && googleSignInAccount != null) {
            SnapshotsClient snapshotsClient = Games.getSnapshotsClient(activity, googleSignInAccount);
            Task<Intent> intentTask = snapshotsClient.getSelectSnapshotIntent(title, allowAddBtn, allowDeleteBtn, maxNumberOfSavedGamesToShow);

            intentTask.addOnSuccessListener(new OnSuccessListener<Intent>() {
                @Override
                public void onSuccess(Intent intent) {
                    activity.startActivityForResult(intent, RC_SAVED_GAMES);
                }
            });
        }
    }

    private void writeSnapshot(Snapshot snapshot, byte[] data, String desc) {
        snapshot.getSnapshotContents().writeBytes(data);
        SnapshotMetadataChange metadataChange = new SnapshotMetadataChange.Builder()
                .setDescription(desc)
                .build();
        GoogleSignInAccount googleSignInAccount = GoogleSignIn.getLastSignedInAccount(activity);
        if (connectionController.isConnected().first && googleSignInAccount != null) {
            SnapshotsClient snapshotsClient = Games.getSnapshotsClient(activity, googleSignInAccount);
            Task<SnapshotMetadata> task = snapshotsClient.commitAndClose(snapshot, metadataChange);
            task.addOnCompleteListener(new OnCompleteListener<SnapshotMetadata>() {
                @Override
                public void onComplete(@NonNull Task<SnapshotMetadata> task) {
                    if (task.isSuccessful()) {
                        godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.SAVED_GAME_SUCCESS, new Object[]{});
                    } else {
                        godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.SAVED_GAME_FAILED, new Object[]{});
                    }
                }
            });
        } else {
            godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.SAVED_GAME_FAILED, new Object[]{});
        }
    }


    protected void saveSnapshot(String gameName, final String dataToSave, final String description) {
        GoogleSignInAccount googleSignInAccount = GoogleSignIn.getLastSignedInAccount(activity);
        if (connectionController.isConnected().first && googleSignInAccount != null) {
            SnapshotsClient snapshotsClient = Games.getSnapshotsClient(activity, googleSignInAccount);
            int conflictResolutionPolicy = SnapshotsClient.RESOLUTION_POLICY_MOST_RECENTLY_MODIFIED;
            
            snapshotsClient.open(gameName, true, conflictResolutionPolicy)
                    .addOnFailureListener(new OnFailureListener() {
                        @Override
                        public void onFailure(@NonNull Exception e) {
                            godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.SAVED_GAME_FAILED, new Object[]{});
                        }
                    }).continueWith(new Continuation<SnapshotsClient.DataOrConflict<Snapshot>, Pair<Snapshot, byte[]>>() {
                @Override
                public Pair<Snapshot, byte[]> then(@NonNull Task<SnapshotsClient.DataOrConflict<Snapshot>> task) throws Exception {
                    SnapshotsClient.DataOrConflict<Snapshot> snapshot = task.getResult();
                    if (snapshot != null && snapshot.getData() != null) {
                        return new Pair<>(snapshot.getData(), toByteArray(dataToSave));
                    }
                    return null;
                }
            }).addOnCompleteListener(new OnCompleteListener<Pair<Snapshot, byte[]>>() {
                @Override
                public void onComplete(@NonNull Task<Pair<Snapshot, byte[]>> task) {
                    if (task.isSuccessful() && task.getResult() != null) {
                        Snapshot snapshot = task.getResult().first;
                        byte[] data = task.getResult().second;
                        writeSnapshot(snapshot, data, description);
                    } else {
                        godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.SAVED_GAME_FAILED, new Object[]{});
                    }
                }
            });
        } else {
            godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.SAVED_GAME_FAILED, new Object[]{});
        }
    }

    void loadSnapshot(String gameName) {
        GoogleSignInAccount googleSignInAccount = GoogleSignIn.getLastSignedInAccount(activity);
        if (connectionController.isConnected().first && googleSignInAccount != null) {
            SnapshotsClient snapshotsClient = Games.getSnapshotsClient(activity, googleSignInAccount);
            int conflictResolutionPolicy = SnapshotsClient.RESOLUTION_POLICY_MOST_RECENTLY_MODIFIED;

            snapshotsClient.open(gameName, true, conflictResolutionPolicy)
                    .addOnFailureListener(new OnFailureListener() {
                        @Override
                        public void onFailure(@NonNull Exception e) {
                            godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.SAVED_GAME_LOAD_FAIL, new Object[]{});
                        }
                    }).continueWith(new Continuation<SnapshotsClient.DataOrConflict<Snapshot>, byte[]>() {
                @Override
                public byte[] then(@NonNull Task<SnapshotsClient.DataOrConflict<Snapshot>> task) throws Exception {
                    SnapshotsClient.DataOrConflict<Snapshot> snapshot = task.getResult();
                    try {
                        if (snapshot != null && snapshot.getData() != null) {
                            return snapshot.getData().getSnapshotContents().readFully();
                        }
                        return null;
                    } catch (IOException e) {
                        Log.e("SavedGamesController", "Error while reading Snapshot.", e);
                    }
                    return null;
                }
            }).addOnCompleteListener(new OnCompleteListener<byte[]>() {
                @Override
                public void onComplete(@NonNull Task<byte[]> task) {
                    if (task.isSuccessful()) {
                        String data = toStringData(task.getResult());
                        godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.SAVED_GAME_LOAD_SUCCESS, new Object[]{data});
                    } else {
                        godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.SAVED_GAME_LOAD_FAIL, new Object[]{});
                    }
                }
            });
        } else {
            godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.SAVED_GAME_LOAD_FAIL, new Object[]{});
        }
    }

    void createNewSnapshot(String currentSaveName) {
        godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.SAVED_GAME_CREATE_SNAPSHOT, new Object[]{currentSaveName});
    }

    private byte[] toByteArray(String data) {
        return data.getBytes();
    }

    private String toStringData(byte[] bytes) {
        return new String(bytes);
    }
}

File: app/src/main/java/io/cgisca/godot/gpgs/savedgames/SavedGamesController.kt
package io.cgisca.godot.gpgs.savedgames

import android.app.Activity
import android.util.Log
import android.util.Pair
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.games.Games
import com.google.android.gms.games.SnapshotsClient
import com.google.android.gms.games.SnapshotsClient.DataOrConflict
import com.google.android.gms.games.snapshot.Snapshot
import com.google.android.gms.games.snapshot.SnapshotMetadataChange
import com.google.android.gms.tasks.Continuation
import io.cgisca.godot.gpgs.ConnectionController
import java.io.IOException

class SavedGamesController(
    private val activity: Activity,
    private val savedGamesListener: SavedGamesListener,
    private val connectionController: ConnectionController
) {

    companion object {
        const val RC_SAVED_GAMES = 9009
        const val NOT_CONNECTED = 1
        const val NOT_SIGNED_IN = 2
        const val WRITE_TASK_FAILED = 3
        const val OPEN_SNAPSHOT_FAILURE = 4
        const val ON_TASK_COMPLETE_FAILURE = 5
    }

    fun showSavedGamesUI(
        title: String,
        allowAddBtn: Boolean,
        allowDeleteBtn: Boolean,
        maxNumberOfSavedGamesToShow: Int
    ) {
        val googleSignInAccount = GoogleSignIn.getLastSignedInAccount(activity)
        if (connectionController.isConnected().first && googleSignInAccount != null) {
            val snapshotsClient = Games.getSnapshotsClient(activity, googleSignInAccount)
            val intentTask = snapshotsClient.getSelectSnapshotIntent(
                title,
                allowAddBtn,
                allowDeleteBtn,
                maxNumberOfSavedGamesToShow
            )
            intentTask.addOnSuccessListener { intent -> activity.startActivityForResult(intent, RC_SAVED_GAMES) }
        }
    }

    private fun writeSnapshot(
        snapshot: Snapshot,
        data: ByteArray?,
        desc: String
    ) {
        snapshot.snapshotContents.writeBytes(data)
        val metadataChange = SnapshotMetadataChange.Builder()
            .setDescription(desc)
            .build()
        val googleSignInAccount = GoogleSignIn.getLastSignedInAccount(activity)
        if (!connectionController.isConnected().first) {
            savedGamesListener.onSavedGameFailed(NOT_CONNECTED)
        }
        else if (googleSignInAccount == null) {
            savedGamesListener.onSavedGameFailed(NOT_SIGNED_IN)
        }
        else {
            val snapshotsClient = Games.getSnapshotsClient(activity, googleSignInAccount)
            val commitTask = snapshotsClient.commitAndClose(snapshot, metadataChange)
            commitTask.addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    savedGamesListener.onSavedGameSuccess()
                } else {
                    savedGamesListener.onSavedGameFailed(WRITE_TASK_FAILED)
                }
            }
        }
    }

    fun saveSnapshot(
        gameName: String,
        dataToSave: String,
        description: String
    ) {
        val googleSignInAccount = GoogleSignIn.getLastSignedInAccount(activity)
        if (! connectionController.isConnected().first) {
            savedGamesListener.onSavedGameFailed(NOT_CONNECTED)
        }
        else if (googleSignInAccount == null) {
            savedGamesListener.onSavedGameFailed(NOT_SIGNED_IN)
        }
        else {
            val snapshotsClient = Games.getSnapshotsClient(activity, googleSignInAccount)
            val conflictResolutionPolicy = SnapshotsClient.RESOLUTION_POLICY_MOST_RECENTLY_MODIFIED
            val createIfNotFound = true
            snapshotsClient.open(gameName, createIfNotFound, conflictResolutionPolicy)
                .addOnFailureListener {
                    savedGamesListener.onSavedGameFailed(OPEN_SNAPSHOT_FAILURE)
                }
                .continueWith<Pair<Snapshot, ByteArray>>(
                    Continuation<DataOrConflict<Snapshot>, Pair<Snapshot, ByteArray>> { task ->
                        val snapshot = task.result
                        snapshot?.data?.let {
                            return@Continuation Pair(it, toByteArray(dataToSave))
                        }
                        null
                    })
                .addOnCompleteListener { task ->
                    if (task.isSuccessful && task.result != null) {
                        val snapshot = task.result!!.first
                        val data = task.result!!.second
                        writeSnapshot(snapshot, data, description)
                    } else {
                        savedGamesListener.onSavedGameFailed(ON_TASK_COMPLETE_FAILURE)
                    }
                }
        }
    }

    fun loadSnapshot(gameName: String) {
        val googleSignInAccount = GoogleSignIn.getLastSignedInAccount(activity)
        if (! connectionController.isConnected().first) {
            savedGamesListener.onSavedGameLoadFailed(NOT_CONNECTED)
        }
        else if (googleSignInAccount == null) {
            savedGamesListener.onSavedGameLoadFailed(NOT_SIGNED_IN)
        }
        else {
            val snapshotsClient = Games.getSnapshotsClient(activity, googleSignInAccount)
            val conflictResolutionPolicy = SnapshotsClient.RESOLUTION_POLICY_MOST_RECENTLY_MODIFIED
            val createIfNotFound = true
            snapshotsClient.open(gameName, createIfNotFound, conflictResolutionPolicy)
                .addOnFailureListener {
                    savedGamesListener.onSavedGameLoadFailed(OPEN_SNAPSHOT_FAILURE)
                }
                .continueWith<ByteArray>(Continuation { task ->
                    val snapshot = task.result
                    try {
                        snapshot?.data?.let {
                            return@Continuation it.snapshotContents.readFully()
                        }
                        return@Continuation null
                    } catch (e: IOException) {
                        Log.e("SavedGamesController", "Error while reading Snapshot.", e)
                    }
                    null
                })
                .addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        task.result?.let {
                            val data = toStringData(it)
                            savedGamesListener.onSavedGameLoadSuccess(data)
                        }
                    } else {
                        savedGamesListener.onSavedGameLoadFailed(ON_TASK_COMPLETE_FAILURE)
                    }
                }
        }
    }

    fun createNewSnapshot(currentSaveName: String) {
        savedGamesListener.onSavedGameCreateSnapshot(currentSaveName)
    }

    private fun toByteArray(data: String): ByteArray {
        return data.toByteArray()
    }

    private fun toStringData(bytes: ByteArray): String {
        return String(bytes)
    }
}